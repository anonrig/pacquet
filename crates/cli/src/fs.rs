use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


