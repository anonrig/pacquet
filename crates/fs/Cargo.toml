[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
