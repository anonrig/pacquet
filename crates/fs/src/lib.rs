mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
