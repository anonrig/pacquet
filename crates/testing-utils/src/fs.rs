use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
