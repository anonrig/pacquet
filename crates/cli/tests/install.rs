use assert_cmd::prelude::*;
use command_extra::CommandExtra;
use pacquet_testing_utils::{
    bin::pacquet_with_temp_cwd,
    fs::{get_all_files, get_all_folders, is_symlink_or_junction},
};
use pipe_trait::Pipe;
use std::fs;

#[test]
fn should_install_dependencies() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "is-odd": "3.0.1",
        },
        "devDependencies": {
            "fast-decode-uri-component": "1.0.1",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Make sure the package is installed");
    assert!(is_symlink_or_junction(&workspace.join("node_modules/is-odd")).unwrap());
    assert!(workspace.join("node_modules/.pacquet/is-odd@3.0.1").exists());

    eprintln!("Make sure it installs direct dependencies");
    assert!(!workspace.join("node_modules/is-number").exists());
    assert!(workspace.join("node_modules/.pacquet/is-number@6.0.0").exists());

    eprintln!("Make sure we install dev-dependencies as well");
    assert!(
        is_symlink_or_junction(&workspace.join("node_modules/fast-decode-uri-component")).unwrap()
    );
    assert!(workspace.join("node_modules/.pacquet/fast-decode-uri-component@1.0.1").is_dir());

    eprintln!("Snapshot");
    let workspace_folders = get_all_folders(&workspace);
    let store_files = get_all_files(&root.path().join("pacquet-store"));
    insta::assert_debug_snapshot!((workspace_folders, store_files));

    drop(root); // cleanup
}

#[test]
fn should_install_exec_files() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "pretty-exec": "0.3.10",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Listing all files in the store...");
    let store_files = root.path().join("pacquet-store").pipe_as_ref(get_all_files);
    let (exec_files, non_exec_files): (Vec<_>, Vec<_>) =
        store_files.iter().partition(|path| path.ends_with("-exec"));

    eprintln!("Snapshot");
    insta::assert_debug_snapshot!(exec_files);

    #[cfg(unix)]
    {
        use pacquet_testing_utils::fs::is_path_executable;
        use pretty_assertions::assert_eq;

        eprintln!("All files that end with '-exec' are executable, others not");
        assert_eq!(
            store_files.iter().partition(|name| {
                root.path().join("pacquet-store").join(name).pipe_as_ref(is_path_executable)
            }),
            (exec_files, non_exec_files),
        );
    }

    #[cfg(windows)]
    let _ = non_exec_files;

    drop(root); // cleanup
}
