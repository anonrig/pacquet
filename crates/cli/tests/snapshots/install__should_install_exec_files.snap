---
source: crates/cli/tests/install.rs
assertion_line: 105
expression: store_files
---
[
    "v3/files/02/f9d952341eade675d0a8f5da14d4277fded62f937242292c2cf5f1bb0eda101b5f803098d64e2a138f4bfe4cdf326cfaf8f6b4015476ab86d25df03ddec731-exec",
    "v3/files/68/b837b4fd7982b081521e51b14a167ccdb272b3c499202449d07d48d78a2fa7bb8652d3a894bb06ac6e3995e5126ee3470f73b5775584940751fce8be468405",
    "v3/files/93/e31fcd9144a604f87d04fa8fb4a058202c579df17fd2b21301e17ac0ae4f3e2bae90ced7ae71b71218bd4a789e355ef7b8bd8899cdb785250071b229503b2e",
    "v3/files/a5/541962367c9c3aaaa8be96312f7efedf7ae638bd051e80573649ca93b4852f4f2749efdce9f19fa7355b3a6d9130eded64e4b3c0efe753622735c5069482cd-index.json",
    "v3/files/bf/0c878f2a4f2ec6f231bb2d414547efb15eb22a673815cba36acfd89633cd4c8affcbdc9378e90ad4d8fc654de921d1f67444be94e6a2f81bda9330be60be93-exec",
    "v3/files/dd/feae8107ea156f46477e8e49e4dfb12365ba9ee122c953db785a5f776f4ff7b91b25378889dab77fb4188f8caaf45cc1c373b01ad326aef20d6a77c24ecefd-index.json",
]
